src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -2837,6 +2838,25 @@ namespace QgsWms
     if ( scaleDenominator > 0 )
       legendModel->setLegendFilterByScale( scaleDenominator );
 
+    // countSymbolFeatures is launched at the createLayerTreeModelLegendNodes
+    // with showFeatureCount property
+    if ( showFeatureCount )
+    {
+      for ( QgsMapLayer *ml : layers )
+      {
+        if ( ml->type() != QgsMapLayer::VectorLayer )
+          continue;
+
+        QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( ml );
+
+        QgsVectorLayerFeatureCounter *counter = vl->countSymbolFeatures();
+        if ( !counter )
+          continue;
+
+        counter->waitForFinished();
```

Comment:
Now if you want to get a bit more performance you can parallelize this:

Loop over all the layers, start a counter for each, collect them in a list and after the loop wait for every entry in the list to be finished. It will then query all the datasources for feature counts at the same time.